test/report/report.cpp

```diff
@@ -0,0 +1,250 @@
+        return do_write_entry_(e, [=](Callback<Error> cb) {
+            if (write_count++ == 0) {
+                cb(MockedError());
+                return;
+            }
+            cb(NoError());
+        });
+    }
+
+    Continuation<Error> close() override {
+        return do_close_([=](Callback<Error> cb) {
+            if (close_count++ == 0) {
+                cb(MockedError());
+                return;
+            }
+            cb(NoError());
+        });
+    }
+
+    int close_count = 0;
+    int open_count = 0;
+    int write_count = 0;
+};
+
+FailingReporter::~FailingReporter() {}
+
+TEST_CASE("The constructor works correctly") {
+    REQUIRE_NOTHROW(Report());
+}
+
+TEST_CASE("The open() method works correctly") {
+    Report report;
+    report.add_reporter(BaseReporter::make());
+    report.open([&](Error err) {
+        REQUIRE(!err);
+        report.open([&](Error err) {
+            REQUIRE(err.code == NoError().code);
+            REQUIRE(err.child_errors.size() == 1);
+            REQUIRE(err.child_errors[0]->code == NoError().code);
+            REQUIRE(err.child_errors[0]->child_errors.size() == 1);
+            REQUIRE(err.child_errors[0]->child_errors[0]->code ==
+                    ReportAlreadyOpenError().code);
+        });
+    });
+}
+
+TEST_CASE("We can retry a partially successful open") {
+    Var<CountedReporter> counted_reporter = CountedReporter::make();
+    Var<FailingReporter> failing_reporter = FailingReporter::make();
+    Report report;
+    report.add_reporter(counted_reporter);
+    report.add_reporter(failing_reporter);
+    report.open([&](Error err) {
+        REQUIRE(err.code == ParallelOperationError().code);
+        REQUIRE(err.child_errors.size() == 2);
+        REQUIRE(err.child_errors[0]->code == NoError().code);
+        REQUIRE(err.child_errors[1]->code == MockedError().code);
+        report.open([&](Error err) {
+            REQUIRE(err.code == NoError().code);
+            REQUIRE(err.child_errors.size() == 2);
+            REQUIRE(err.child_errors[0]->code == NoError().code);
+            REQUIRE(err.child_errors[0]->child_errors.size() == 1);
+            REQUIRE(err.child_errors[0]->child_errors[0]->code ==
+                    ReportAlreadyOpenError().code);
+            REQUIRE(err.child_errors[1]->code == NoError().code);
+        });
+    });
+    REQUIRE(counted_reporter->open_count == 1);
+    REQUIRE(failing_reporter->open_count == 2);
+}
+
+TEST_CASE("The write_entry() method works correctly") {
+    Report report;
+    report.add_reporter(BaseReporter::make());
+    Entry entry;
+    report.write_entry(entry, [&](Error err) {
+        REQUIRE(err.code == ParallelOperationError().code);
+        REQUIRE(err.child_errors.size() == 1);
+        REQUIRE(err.child_errors[0]->code == ReportNotOpenError().code);
+        report.open([&](Error err) {
+            REQUIRE(!err);
+            report.write_entry(entry, [&](Error err) {
```

Comment:
xxx: because I've added caching, this test is now actually not testing what I would like to test